### DISCUSSION TITLE
How to know when required field was not provided?

### DISCUSSION THREAD

**Discussion by jeanlst** [pydantic/pydantic#3729] [Question] [answered]
Hello,

So I have a required field in my BaseConfig that in case I forgot to put it in the `.env` instead of throwing an error saying missing value, I wanna use a default generated value and print something like "You forgot this field. Using this generated one blabla". How can I do this? When I use a validator with `pre=True, always=True` I have no idea if the value is coming from the default or from the `.env`

**Comment by uriyyo** [comment, marked as answer]
You can use custom `default_factory`:
```py
from typing import Any

from pydantic import BaseSettings, Field


def _generate_default() -> Any:
    value = "1234"  # generate default value

    print(f"You forgot this field. Using this generated one {value}")

    return value


class Settings(BaseSettings):
    VALUE: str = Field(default_factory=_generate_default)


Settings()  # prints 
Settings(VALUE="123")  # didn't print nothing
```
